README.md
# JobRangers

JobRangers is a powerful tool for those looking to create impressive resumes and find relevant job opportunities. Our vision is to empower users to present their skills and experiences in the best possible light and connect them with the perfect match for their personalities. Our ultimate goal is to help job seekers land their dream job by providing them with the tools and resources they need to succeed. Whether users are just starting their careers or looking for a new opportunity, JobRangers is the app that will give them a competitive edge in the current job market.

## Contributors

- Sam Adams
- Max Eaton
- Gerardo Ferrigno
- Grant Hargrave
- Prajwal Nepal
- Sang Shim

## Built with

- Docker - containerization platform
- PostgreSQL - relational database management system
- Node.js - JavaScript runtime environment
- EJS - templating language for HTML
- CSS - styling language for web pages

## Pages

- Home: This page redirects to other pages and has an "About Us" section.
- Discover: This page lists jobs based on user skills.
- Info: This page has fields to input user data and generates a PDF of a resume.

## Installation and Deployment

To install JobRangers, run the following command:

```
git clone https://github.com/CSCI-3308-Team-03/JobRangers.git
```

To deploy locally, make sure you have Docker installed and run the following command:

```
docker-compose up
```

To reset the database, run:

```
docker-compose down --volumes
```

After deployment, JobRangers can be accessed at `http://localhost:3000`.
	
## Azure Deployment
The app is also deployed on Azure at `http://recitation-12-team-03.eastus.cloudapp.azure.com`.

## Communication Plan
Our team plans to utilize Discord to discuss meeting times, coordinate person-to-person workload, and complete future project milestones.

## Meeting Pla
The TA meeting will be held every Monday at 3:45 PM. The scrum meeting will occur on Zoom from 4:00 PM to 5:00 PM on Monday as well. This will allow the members on the team to receive inputs directly from the TAs and move into development without any delay. 
	
## API Information
Using a REST API from an official website of the United States Government which stores potential national job listings. We will be using the api/search/ endpoint which returns an array of objects containing jobs along with their respective attributes and descriptions.

 - Website <https://developer.usajobs.gov/>
 - API KEY <QYaokDz2ueAHu3iPkUCh8Zn7wBR11Hl0l7ruwzfGJ8U=>
 - Email <granthargrav@gmail.com>



milestoneSubmissions/ReleaseNotes.txt
Release Notes:

0.1.0
	Login and register
	Staging for ejs pages
0.1.1
	Created sql databases
0.1.2
	Created UATs
	Created demo user
0.1.3
	Created pdf export feature
0.1.4
	Created API calls for external API
0.1.5
	Added partials to pages
0.1.6
	Fixed minor bugs
0.2.0
	Integrated UI and functional aspects
0.2.1
	Fixed bugs 
0.2.2
	Deployed page to cloud


src/test/server.spec.js
const server = require('../index');
// Importing libraries

const chai = require('chai');
const chaiHttp = require('chai-http');
chai.should();
chai.use(chaiHttp);
const {assert, expect} = chai;

describe('Register!', () => {
  it('positive : /register', done => {
      chai
        .request(server)
        .post('/register')
        .send({username: 'test123', password: 'test234'})
        .end((err, res) => {
          expect(res).to.have.status(200);
          done();
        });
    })
  it('Negative : /register. Checking invalid name', done => {
    chai
      .request(server)
      .post('/register')
      .send({username: '', password: ''})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
});

describe('login!', () => {
  it('positive : /login', done => {
      chai
        .request(server)
        .post('/login')
        .send({username: 'test123', password: 'test234'})
        .end((err, res) => {
          expect(res).to.have.status(200);
          done();
        });
    })
  it('Negative : /login. Checking invalid name', done => {
    chai
      .request(server)
      .post('/login')
      .send({id: '5', username: '', password: ''})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
});

describe('Add General', () => {
  it('positive : /info/addGeneral', done => {
    chai
      .request(server)
      .post('/info/addGeneral')
      .send({username: 'test123', firstname: 'Bob', lastname: 'Smith', dob: '1995-02-02', email: '123@gmail.com', linkedin: 'bSmith@linkedin.com', github: 'bsmith@github.com'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/addGeneral. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/addGeneral')
      .send({username: 'test123', firstname: 6, lastname: 'Smith', dob: 45, email: '123@gmail.com', linkedin: 'bSmith@linkedin.com', github: 'bsmith@github.com'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Update General', () => {
  it('positive : /info/updateGeneral', done => {
    chai
      .request(server)
      .post('/info/updateGeneral')
      .send({username: 'test123', firstname: 'Bob', lastname: 'Smith', dob: '1995-02-02', email: 'bobbyBoiFre$$$h@gmail.com', linkedin: 'bSmith@linkedin.com', github: 'bsmith@github.com'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/updateGeneral. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/updateGeneral')
      .send({username: 'test123', firstname: 1, lastname: 'Smith', dob: 1, email: '123@gmail.com', linkedin: 'bSmith@linkedin.com', github: 'bsmith@github.com'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Add Education', () => {
  it('positive : /info/addEducation', done => {
    chai
      .request(server)
      .post('/info/addEducation')
      .send({username: 'test123', school: 'East High School', degree: 'GED', focus: 'Keepin it real', startdate: '2007-12-12', enddate: '2011-05-05', description: 'You know what it is'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/addEducation. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/addEducation')
      .send({username: 'test123', school: 'East High School', degree: 55, focus: 'Keepin it real', startdate: '2007-32-32', enddate: '2011-05-05', description: 'You know what it is'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Update Education', () => {
  it('positive : /info/updateEducation', done => {
    chai
      .request(server)
      .post('/info/updateEducation')
      .send({username: 'test123', school: 'West High School', degree: 'GED', focus: 'Keepin it real', startdate: '2007-12-12', enddate: '2011-05-05', description: 'You know what it is'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/updateEducation. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/updateEducation')
      .send({username: 'test123', school: 'East High School', degree: 55, focus: 'Keepin it real', startdate: '2007-32-32', enddate: '2011-05-05', description: 'You know what it is'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Add Experience', () => {
  it('positive : /info/addExperience', done => {
    chai
      .request(server)
      .post('/info/addExperience')
      .send({username: 'test123', organization: 'The Haus', title: 'Boss', startdate: '2011-11-11', enddate: '2014-11-11', description: 'Hahahahaha'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/addExperience. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/addExperience')
      .send({username: 'test123', organization: 555, title: 'Boss', startdate: '2011-11-11', enddate: '2014-11-11', description: 'Hahahahaha'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Update Experience', () => {
  it('positive : /info/updateExperience', done => {
    chai
      .request(server)
      .post('/info/updateExperience')
      .send({username: 'test123', organization: 'The House', title: 'Peasant', startdate: '2011-11-11', enddate: '2014-11-11', description: 'Hahahahaha'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/updateExperience. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/updateExperience')
      .send({username: 'test123', organization: 444, title: 'Boss', startdate: '2011-11-11', enddate: '2014-11-11', description: 'Hahahahaha'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Add Skill', () => {
  it('positive : /info/addSkill', done => {
    chai
      .request(server)
      .post('/info/addSkill')
      .send({username: 'test123', skill: 'Hustle Pool'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/addSkill. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/addSkill')
      .send({username: '1', skill: 55})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Update Skill', () => {
  it('positive : /info/updateSkill', done => {
    chai
      .request(server)
      .post('/info/updateSkill')
      .send({username: 'test123', skill: 'Play Pool'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/updateSkill. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/updateSkill')
      .send({username: 'test123', skill: 44})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Add language', () => {
  it('positive : /info/addLanguage', done => {
    chai
      .request(server)
      .post('/info/addLanguage')
      .send({username: 'test123', language: 'Italian', proficiency: 'Fluent'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/addLanguage. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/addLanguage')
      .send({username: 'test123', language: 14, proficiency: 'Fluent'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Update language', () => {
  it('positive : /info/updateLanguage', done => {
    chai
      .request(server)
      .post('/info/updateLanguage')
      .send({username: 'test123', language: 'Italian', proficiency: 'Begginer'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/updateLanguage. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/updateLanguage')
      .send({username: 'test123', language: 12, proficiency: 'Fluent'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Add Location', () => {
  it('positive : /info/addLocation', done => {
    chai
      .request(server)
      .post('/info/addLocation')
      .send({username: 'test123', country: 'United States', city: 'Miami'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/addLocation. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/addLocation')
      .send({username: 'test123', country: 14, city: 'Miami'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

describe('Update Location', () => {
  it('positive : /info/updateLocation', done => {
    chai
      .request(server)
      .post('/info/updateLocation')
      .send({username: 'test123', country: 'United States', city: 'New York'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  })
  it('Negative : /info/updateLocation. Checking invalid input', done => {
    chai
      .request(server)
      .post('/info/updateLocation')
      .send({username: 'test123', country: 14, city: 'Miami'})
      .end((err, res) => {
        expect(res).to.have.status(200);
        done();
      });
  });
}); 

src/index.js
// *****************************************************
// <!-- Section 1 : Import Dependencies -->
// *****************************************************

const express = require('express'); // To build an application server or API
const app = express();
const pgp = require('pg-promise')(); // To connect to the Postgres DB from the node server
const bodyParser = require('body-parser');
const session = require('express-session'); // To set the session object. To store or access session data, use the `req.session`, which is (generally) serialized as JSON by the store.
const bcrypt = require('bcrypt'); //  To hash passwords
const axios = require('axios'); // To make HTTP requests from our server. We'll learn more about it in Part B.

// *****************************************************
// <!-- Section 2 : Connect to DB -->
// *****************************************************

// database configuration
const dbConfig = {
  host: 'db', // the database server
  port: 5432, // the database port
  database: process.env.POSTGRES_DB, // the database name
  user: process.env.POSTGRES_USER, // the user account to connect with
  password: process.env.POSTGRES_PASSWORD, // the password of the user account
};

const db = pgp(dbConfig);

// test your database
db.connect()
  .then(obj => {
    console.log('Database connection successful'); // you can view this message in the docker compose logs
    obj.done(); // success, release the connection;
  })
  .catch(error => {
    console.log('ERROR:', error.message || error);
  });

// *****************************************************
// <!-- Section 3 : App Settings -->
// *****************************************************

app.set('view engine', 'ejs'); // set the view engine to EJS
app.use(bodyParser.json()); // specify the usage of JSON for parsing request body.

// initialize session variables
app.use(
  session({
    secret: process.env.SESSION_SECRET,
    saveUninitialized: false,
    resave: false,
  })
);

app.use(
  bodyParser.urlencoded({
    extended: true,
  })
);

const user = {
  username: undefined,
  password: undefined
}

// *****************************************************
// <!-- Section 4 : API Routes -->
// *****************************************************

app.get('/', (req, res) => {
    res.redirect('/login');
});
  
app.get('/register', (req, res) => {
  res.render('pages/register', {unauth: true,});
});

app.post('/register', async (req, res) => {
  const username = req.body.username;
  const password = await bcrypt.hash(req.body.password, 10);
  const insert = `INSERT INTO users (username, password) VALUES ('${username}', '${password}');`;

  db.task('get-everything', task => {
      return task.any(insert);
  })
      .then(data => {
        res.redirect("pages/login");
      })
      .catch(err => {
        console.log(err);
        res.locals.message = 'Username already exists';
        res.render("pages/register", {unauth: true,});
      });
});

app.get('/login', (req, res) => {
    res.render("pages/login", {unauth: true,});
});

app.post('/login', async (req, res) => {
    const access = `SELECT * FROM users WHERE username = '${req.body.username}';`;

    db.task('get-everything', async task => {
      return task.any(access);
    })
      .then(async data => {
        if (await bcrypt.compare(req.body.password, data[0].password)) {
          user.username = data[0].username;
          user.password = data[0].password;
          req.session.user = user;
          req.session.save();

          res.redirect('/home');
        } else {
          res.locals.message = 'Incorrect username or password';
          res.status(200).render("pages/login", {unauth: true,});
        }
      })
      .catch(err => {
        res.locals.message = 'Incorrect username or password';
        res.status(200).render("pages/login", {unauth: true,});
      });
});

const auth = (req, res, next) => {
  if (!req.session.user) {
    return res.redirect('/login');
  }
  next();
};

app.use(auth);

app.get('/logout', (req, res) => {
  user.username = undefined;
  user.password = undefined;
  req.session.user = user;
  req.session.save();
  res.locals.message = 'Logged out successfully';
  res.render("pages/login", {unauth: true,});
});

app.get('/home', (req, res) => {
  res.render("pages/home");
});

const getData = async () => {
  const getGeneral = `SELECT * FROM general WHERE username = '${user.username}'`;
  const getEducations = `SELECT * FROM educations WHERE username = '${user.username}'`;
  const getExperiences = `SELECT * FROM experiences WHERE username = '${user.username}'`;
  const getSkills = `SELECT * FROM skills WHERE username = '${user.username}'`;
  const getLanguages = `SELECT * FROM languages WHERE username = '${user.username}'`;
  const getLocations = `SELECT * FROM locations WHERE username = '${user.username}'`;

  try {
    const data = await db.task('get-everything', task => {
      return task.batch([task.any(getGeneral), task.any(getEducations), task.any(getExperiences), task.any(getSkills), task.any(getLanguages), task.any(getLocations)]);
    });
    return data;
  } 
  catch (error) {
    return [];
  }
};

app.get('/info', async (req, res) => {
  return res.render("pages/info", {
    data: await getData(),
  });
});

app.post('/info/addGeneral', async (req, res) => {
  const add = `INSERT INTO general (firstname, lastname, dob, email, linkedin, github, username) SELECT '${req.body.firstname}', '${req.body.lastname}', '${req.body.dob}', '${req.body.email}', '${req.body.linkedin}', '${req.body.github}', '${user.username}' WHERE NOT EXISTS (SELECT * FROM general WHERE username = '${user.username}');`;
  db.task('get-everything', task => {
    return task.any(add);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to add general information",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to add general information",
      });
    });
});

app.post('/info/editGeneral', async (req, res) => {
  const edit = `SELECT * FROM general WHERE id = '${req.body.id}';`;
  db.task('get-everything', task => {
    return task.one(edit);
  })
    .then(async data => {
      res.render("pages/edit", {
        data: [[data], null, null, null, null, null],
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to edit general information",
      });
    });
});

app.post('/info/updateGeneral', async (req, res) => {
  const update = `UPDATE general SET firstname='${req.body.firstname}', lastname='${req.body.lastname}', dob='${req.body.dob}', email='${req.body.email}', linkedin='${req.body.linkedin}', github='${req.body.github}' WHERE id = ${req.body.id};`;
  db.task('get-everything', task => {
    return task.any(update);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to update general information",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to update general information",
      });
    });
});

app.post('/info/addEducation', async (req, res) => {
  const add = `INSERT INTO educations (school, degree, focus, startdate, enddate, description, username) SELECT '${req.body.school}', '${req.body.degree}', '${req.body.focus}', '${req.body.startdate}', '${req.body.enddate}', '${req.body.description}', '${user.username}' WHERE NOT EXISTS (SELECT * FROM educations WHERE school = '${req.body.school}' AND degree = '${req.body.degree}' AND focus = '${req.body.focus}' AND username = '${user.username}');`;
  db.task('get-everything', task => {
    return task.any(add);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to add education",
      });
    })
    .catch(async err => {
      console.log(err);
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to add education",
      });
    });
});

app.post('/info/editEducation', async (req, res) => {
  const edit = `SELECT * FROM educations WHERE id = '${req.body.id}';`;
  db.task('get-everything', task => {
    return task.any(edit);
  })
    .then(async data => {
      res.render("pages/edit", {
        data: [null, [data], null, null, null, null],
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to edit education information",
      });
    });
});

app.post('/info/updateEducation', async (req, res) => {
  const update = `UPDATE educations SET school='${req.body.school}', degree='${req.body.degree}', focus='${req.body.focus}', startdate='${req.body.startdate}', enddate='${req.body.enddate}', description='${req.body.description}' WHERE id = ${req.body.id};`;
  db.task('get-everything', task => {
    return task.any(update);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to update education information",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to update education information",
      });
    });
});

app.post('/info/rmEducation', (req, res) => {
  const rm = `DELETE FROM educations WHERE id = '${req.body.id}';`;

  db.task('get-everything', task => {
    return task.any(rm);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to delete education",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to delete education",
      });
    });
});

app.post('/info/addExperience', async (req, res) => {
  const add = `INSERT INTO experiences (organization, title, startdate, enddate, description, username) SELECT '${req.body.organization}', '${req.body.title}', '${req.body.startdate}', '${req.body.enddate}', '${req.body.description}','${user.username}' WHERE NOT EXISTS (SELECT * FROM experiences WHERE organization = '${req.body.organization}' AND title = '${req.body.title}' AND username = '${user.username}');`;

  db.task('get-everything', task => {
    return task.any(add);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to add experience",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to add experience",
      });
    });
});

app.post('/info/editExperience', async (req, res) => {
  const edit = `SELECT * FROM experiences WHERE id = '${req.body.id}';`;
  db.task('get-everything', task => {
    return task.one(edit);
  })
    .then(async data => {
      res.render("pages/edit", {
        data: [null, null, [data], null, null, null],
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to edit experience information",
      });
    });
});

app.post('/info/updateExperience', async (req, res) => {
  const update = `UPDATE experiences SET organization='${req.body.organization}', title='${req.body.title}', startdate='${req.body.startdate}', enddate='${req.body.enddate}', description='${req.body.description}' WHERE id = ${req.body.id};`;
  db.task('get-everything', task => {
    return task.any(update);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to update experience information",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to update experience information",
      });
    });
});

app.post('/info/rmExperience', (req, res) => {
  const rm = `DELETE FROM experiences WHERE id = '${req.body.id}';`;

  db.task('get-everything', task => {
    return task.any(rm);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to delete experience",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to delete experience",
      });
    });
});

app.post('/info/addSkill', async (req, res) => {
  const add = `INSERT INTO skills (skill, username) SELECT '${req.body.skill}', '${user.username}' WHERE NOT EXISTS (SELECT * FROM skills WHERE skill = '${req.body.skill}' AND username = '${user.username}');`;

  db.task('get-everything', task => {
    return task.any(add);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to add skill",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to add skill",
      });
    });
});

app.post('/info/editSkill', async (req, res) => {
  const edit = `SELECT * FROM skills WHERE id = '${req.body.id}';`;
  db.task('get-everything', task => {
    return task.one(edit);
  })
    .then(async data => {
      res.render("pages/edit", {
        data: [null, null, null, [data], null, null],
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to edit skill information",
      });
    });
});

app.post('/info/updateSkill', async (req, res) => {
  const update = `UPDATE skills SET skill='${req.body.skill}' WHERE id = ${req.body.id};`;
  db.task('get-everything', task => {
    return task.any(update);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to update skill information",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to update skill information",
      });
    });
});

app.post('/info/rmSkill', async (req, res) => {
  const rm = `DELETE FROM skills WHERE id = '${req.body.id}';`;

  db.task('get-everything', task => {
    return task.any(rm);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to delete skill",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to delete skill",
      });
    });
});

app.post('/info/addLanguage', async (req, res) => {
  const add = `INSERT INTO languages (language, proficiency, username) SELECT '${req.body.language}', '${req.body.proficiency}', '${user.username}' WHERE NOT EXISTS (SELECT * FROM languages WHERE language = '${req.body.language}' AND username = '${user.username}');`;

  db.task('get-everything', task => {
    return task.any(add);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to add language",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to add language",
      });
    });
});

app.post('/info/editLanguage', async (req, res) => {
  const edit = `SELECT * FROM languages WHERE id = '${req.body.id}';`;
  db.task('get-everything', task => {
    return task.one(edit);
  })
    .then(async data => {
      res.render("pages/edit", {
        data: [null, null, null, null, [data], null],
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to edit language information",
      });
    });
});

app.post('/info/updateLanguage', async (req, res) => {
  const update = `UPDATE languages SET language='${req.body.language}', proficiency='${req.body.proficiency}' WHERE id = ${req.body.id};`;
  db.task('get-everything', task => {
    return task.any(update);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to update language information",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to update language information",
      });
    });
});

app.post('/info/rmLanguage', async (req, res) => {
  const rm = `DELETE FROM languages WHERE id = '${req.body.id}';`;

  db.task('get-everything', task => {
    return task.any(rm);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to delete language",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to delete language",
      });
    });  
});

app.post('/info/addLocation', async (req, res) => {
  const add = `INSERT INTO locations (country, city, username) SELECT '${req.body.country}', '${req.body.city}', '${user.username}' WHERE NOT EXISTS (SELECT * FROM locations WHERE country = '${req.body.country}' AND city = '${req.body.city}' AND username = '${user.username}');`;

  db.task('get-everything', task => {
    return task.any(add);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to add location",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to add location",
      });
    });
});

app.post('/info/editLocation', async (req, res) => {
  const edit = `SELECT * FROM locations WHERE id = '${req.body.id}';`;
  db.task('get-everything', task => {
    return task.one(edit);
  })
    .then(async data => {
      res.render("pages/edit", {
        data: [null, null, null, null, null, [data]],
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to edit location information",
      });
    });
});

app.post('/info/updateLocation', async (req, res) => {
  const update = `UPDATE locations SET country='${req.body.country}', city='${req.body.city}' WHERE id = ${req.body.id};`;
  db.task('get-everything', task => {
    return task.any(update);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to update location information",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to update location information",
      });
    });
});

app.post('/info/rmLocation', async (req, res) => {
  const rm = `DELETE FROM locations WHERE id = '${req.body.id}';`;

  db.task('get-everything', task => {
    return task.any(rm);
  })
    .then(async data => {
      res.render("pages/info", {
        data: await getData(),
        message: "Succeeded to delete location",
      });
    })
    .catch(async err => {
      res.render("pages/info", {
        data: await getData(),
        message: "Failed to delete location",
      });
    });
});

app.get('/template', async (req,res) => {
  return res.render("pages/resume_1", {
    result: await getData(),
  });
});

app.get('/home', (req, res) => {
  res.render("pages/home")
});






app.get('/discover' , async (req,res) =>{
  const getSkills = `SELECT * FROM skills WHERE username = '${user.username}'`;
  db.any(getSkills)
  .then(function(skills){
     console.log(skills)
    // console.log(skills[0].skill)
   // console.log(skills[0][1])
    // console.log("getSkills should hav ebeen printed")
    axios({      
        url: `https://data.usajobs.gov/api/search?Keyword=${skills[0].skill}`,      
        method: 'GET',      
        headers: {          
            "Host": process.env.HOST,          
            "User-Agent": process.env.USERAGENT,          
            "Authorization-Key": process.env.AUTHKEY      
        }  
    })
    .then(results => {
      console.log(results) 
      data =  results.data.SearchResult.SearchResultItems; 
     // console.log(data)
      res.render('pages/discover',{data:data})

    })
    .catch(error => {
      console.log(error)
      res.render('pages/home',{message:"Something went wrong"});
    });
  })
  .catch(error =>{
    console.log(error)
    res.render('pages/home',{message:"Something went wrong"});
  })
  
  
});

app.use(express.static("public"));

// *****************************************************
// <!-- Section 5 : Start Server-->
// *****************************************************
// starting the server and keeping the connection open to listen for more requests
module.exports = app.listen(3000);
console.log('Server is listening on port 3000');
